RFD 0094: Kubernetes node joining (#17905)
gravitational/teleport: rfd/0094-kubernetes-node-joining.md

  `TokenReview` API
- The proxies are part of the cluster

#### Kubernetes tokens

All Kubernetes tokens contain the service account name and its groups.

Versions 1.20 and above ([bound
tokens](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/1205-bound-service-account-tokens/README.md#boundserviceaccounttokenvolume-1)):
- tokens are signed per pod
- tokens are automatically renewed by the kubelet
- token TTL is configurable in the `PodSpec`
- tokens are revoked when the pod is deleted

Versions prior 1.20:
- tokens are signed per service-account
- tokens are long-lived
- tokens are revoked when the service account is deleted

Tokens are JWTs, Kubernetes JWKS can be manually fetched and JWT validation can
be done offline.  Kubernetes also exposes a `TokenReview` API to validate
tokens. This API performs more checks, like revocation, UUID checking pod and
service account, ... This API is the sanest way to validate Kubernetes token.

#### Teleport join tokens

The teleport resource would look like:

```yaml
kind: token
version: v2
metadata:
  name: proxy-token
  expires: "3000-01-01T00:00:00Z"

> spec:

Reviewer: We should aim for consistency with the CircleCI and GitHub work - and pull the configuration for Kubernetes joining into a separate block rather than reusing the overloaded `spec.allow` e.g

```yaml
spec:
  k8s:
    kube_api_server: ""
    allow:
    - service_account: "my-namespace:my-service-account"
```

I'd also suggest ( as shown ) that we allow the URL of a Kubernetes API server to be directly configured. When this value is omitted, the Auth Server should try and detect this from its environment, but by making it configurable, we make it possible for automatic joining from Kubernetes clusters that the auth server is not within.

Author: I think we want to rely on Kubernetes to validate the tokens instead of using the OIDC discovery endpoint and validating this on our own.  Kubernetes does much more than just checking validity, audience and expiry and we cannot benefit from that without valid Kubernetes credentials.

In this context, making the apiserver configurable makes sense, but this would imply:
- having a way to trust the remote Kubernetes certs. Either adding them to the trust store or passing them in the resource.
- a way to pass Kubernetes credentials (I was planning on only using the in-cluster kube client because it made everything easy)

We cannot rely on env vars for this because this would not work if the user create different tokens targeting distinct Kubernetes clusters. We would either have to extend the token definition to contain a working kubeconfig, or to have the user provide kubeconfigs file on all auth nodes. We might also encounter extra issues with cloud-provider-specific auth methods.

I totally see the value of being able top trust additional clusters though. If you're OK I suggest we mention this as a possible next step in the RFD but keep the "remote apiserver" feature out of the first implementation.

Reviewer: Awesome, I wasn't sure how much additional work the special k8s API offered on token validation, but it definitely sounds like its bringing value.

I think I definitely agree w/ keeping this as a next step - just useful to make sure we record whats in and out of scope.

  roles: ["Proxy"]
  join_method: kube
  allow:
  - service_account: "my-namespace:my-service-account"
```

The `allow[*].service_account`check should match exactly the `User` field of
the Kubernetes token stripped of the `system:serviceaccounts:` prefix. Checking
if the `system:serviceaccounts` prefix is present ensures no Kubernetes user
tokens can be used to join the Teleport cluster.

Supporting this token kind does not require changes to the
`RegisterUsingTokenRequest` type, the Kubernetes token should be passed through

> the `Token` field and the Teleport token name should be passed in the `TokenID`

Reviewer: ```suggestion
the `IDToken` field and the Teleport token name should be passed in the `Token`
```

I believe you currently have these reversed, typically `Token` has been used to specify the ProvisionToken that is being used for joining.

---
authors: Hugo Hervieux (hugo.hervieux@goteleport.com)
state: draft
---

# RFD 0094 - Kubernetes node joining

## Required approvers

* Engineering: @r0mant && @gus
* Product: @klizhentas && @xinding33
* Security: @reedloden

## What

This RFD discusses various methods for Teleport proxies to join a Teleport
cluster in a same-cluster Kubernetes context and covers implementation details
of the chosen approach.

## Why

We want to refactor the `teleport-cluster`  Helm chart to be a reference
delivery method for Teleport. This involves deploying separately Teleport auth
and proxy nodes. Proxy nodes have to register against the Teleport auth nodes
without user interaction and without compromising the cluster security (no
long-lived tokens).  Proxy nodes can be automatically created or deleted (by a
Kubernetes HPA for example).

User Motivation: 

- As new Teleport adopter, I want to deploy a working, secure and highly
  available Teleport cluster on Kubernetes, with a single Helm command and as
  few values as possible so that I can start using Teleport as fast as possible.
- As a Teleport power user, I want to easily join my custom sets of Teleport

>   nodes on Kubernetes so that I have less join automation to build and maintain.

Reviewer 2: I don't know if deploying nodes as in "SSH nodes" in the same Kube cluster has much of a value tbh since we have Kube Access, I would maybe say "Teleport services like Machine ID or plugins".

Author: I was thinking about kubernetes_service, app_service, db_service, discovery_service, windows_desktop_service. I'll explicit the list.

Author: done

    (we are due to implement tokens anyway to reach parity between the operator
    and the Terraform provider)
  - this can be done through a bootstrap resource on cluster creation
- Kubernetes only supports (read enabled by default) short-lived renewed token
  after 1.20 (it depends on the Kubernetes distribution, but starting with 1.20
  everyone must support it). For prior versions, users might have to use the legacy
  non-bound tokens produced by Kubernetes.

#### Approach 2 - MachineID-generated certs

In this approach, proxy certificates are delivered by MachineID.

```mermaid
flowchart LR
    cluster[Teleport cluster]
    tBot
    stateA 
    podA
    stateB
    podB

    subgraph statefulset/deployment
    podA
    podB
    end

    cluster <-->|part of cluster| tBot
    tBot --> stateA
    tBot --> stateB
    stateA --- |uses| podA
    stateB --- |uses| podB

```

This approach brings the following challenges:
- each proxy requires multiple certs (one with the default teleport node role,
  the other with the proxy role)
- teleport does not support joining from certs (unless we manually build a
  valid state)

> - tBot has to dynamically discover how many pods want certs and which certs

Reviewer: ```suggestion
- `tbot` has to dynamically discover how many pods want certs and which certs
```

  they need (we don't have a streaming destination API yet)
- pods should have a way to access their specific state:
    - we can't mount a different secret in each pod
    - it can be done through admission webhooks
    - pods can read secrets (this works OK for Teleport nodes, but cannot be
      generalized with other workloads like access plugin because this requires
      being able to read all certs from the namespace)
    - pods can request a cert to tBot over HTTP

> - tBot has to validate the identity of the pod requesting the certificate

Reviewer: ```suggestion
- `tbot` has to validate the identity of the pod requesting the certificate
```

> - tBot has to join the cluster initially, which is almost the same problem that

Reviewer: ```suggestion
- `tbot` has to join the cluster initially, which is almost the same problem that
```

  we're trying to address (almost because we have less constraints due to

>   autoscaling). The easiest approach would be to bundle tBot with auth nodes and

Reviewer: ```suggestion
  autoscaling). The easiest approach would be to bundle `tbot` with auth nodes and
```

```

The operator internally relies on MachineID to get its identity, this solution
is basically an automated version of [MachineID token
generation](https://github.com/gravitational/teleport/discussions/15404).

This approach brings the following challenges:
- the operator becomes mandatory, which might not be a possibility for certain
  users (because of new additional RBAC)
- the operator enters the critical path, if it is not available no pod can start
- pods started before the mutation webhook is added will not get their token
- there is a race condition if the proxy pod does not start before expiry, the
  failure mode is a pod stuck in crashloop and the only way out is to delete
  the pod. This is particularly likely to happen on clusters with node
  autoscaling or capacity issues.

### Choosing the approach

The second approach tends to bring more issues than it solves. Even if we added
MachineID support for Teleport nodes, this would not be used by customers with
their Teleport workload.

The third approach adds new components in the critical path and new complex
failure modes.

The first approach seems the more robust, leverages previous work with

> joinTokens and provides very solid security guarantees for Kubernetes 1.20+.

Reviewer: ```suggestion
ProvisionTokens and provides very solid security guarantees for Kubernetes 1.20+.
```

I think using `ProvisionToken` is more specific terminology (this is the resource name in Teleport) here and avoids conflation with other "tokens" - we have so many tokens 😱

by design: human operators and the CI/CD system should never be in contact with
the token. Legacy tokens are mostly a user experience improvement rather than a
security improvement.

The Teleport Helm charts should always favor using bound tokens. Kubernetes
support for bound tokens must be detected at the deploy time and the chart must
adapt its behaviour to maximize security based on Kubernetes capabilities.

To mitigate token stealing or replaying MiTM attack, the joining
Teleport nodes should be configured with the ca_pins, as for regular tokens.

Introducing the feature does not add new critical components that could cause
availability issues. Creating new pods already requires a healthy apiserver.
If the apiserver were to fail, this would not affect already joined proxies.

#### Implications of adding the feature

While joining Teleport proxy nodes is the main motivation of this proposal,
other Teleport nodes would also benefit from this additional join method.

Deploying nodes in the same Kubernetes cluster will become easier. This can be
used to join all kinds of Teleport service (kubernetes, app, db, windows

> desktop, discovery), but also to join `tBot` and offer an easier MachineID

Reviewer: ```suggestion
desktop, discovery), but also to join `tbot` and offer an easier Machine ID
```

integration in Kubernetes. This would pave the way for a better experience for
users hosting access plugins next to the Teleport control plane.

```mermaid
graph LR
    subgraph cluster helm chart
    cluster
    tBot

    subgraph accessPlugin chart
    accessPlugin
    end

    cluster --- |joined using the 'kube' method| tBot
    tBot --> |store certs| secret
    secret --- |mounted| accessPlugin

    accessPlugin --> |connects with its MachineID certs| cluster
    end
```

Joining nodes from outside the Kubernetes cluster would imply generating and
exporting a Kubernetes token from the service account. While this pattern is
feasible, it does not provide much added value compared to generating a static

> token from Teleport directly.

Reviewer 2: Would it make sense to provide more details on how it would work, maybe in a "future work" section? I just feel like having a "Kubernetes" join method and not supporting agents joining from other Kube clusters is strange. Like, if you run everything on Kubernetes, then you'd probably want to use the same join method even if your stuff is in different clusters? For comparison, IAM join method for example supports nodes joining from different AWS accounts.

Author: It should work the same way, except Teleport uses the kubeconfig from the resource instead of the `inCluster` config.

My biggest concern is that managed kubernetes hostings like GKE, EKS require more than a kubeconfig to interact with them. Suddenly we would have to support all the ways to pass service accounts to Teleport + support each cloud's auth provider in the go code. The spec would look like:

```yaml
kind: token
version: v2
metadata:
  name: proxy-token
  expires: "3000-01-01T00:00:00Z"
spec:
  roles: ["Proxy"]
  k8s:
    allow:
    - service_account: "my-namespace:my-service-account"
      authType: "kubeconfig|gcp|aws|..."
      kubeconfig:
        certificate-authority-data: base64
        server: http://1.2.3.4
        # either set token or user & client-*
        token: my-secret-token
        user: client-sa-name
        client-certificate-data: base64
        client-key-data: base64
      gcp:
        # optional SA account key/name because you can use ambiant credentials or workload identity
        service-account-key: base64
        service-account-name: "foo@sa-project.iam.gserviceaccount.com"
        # mandatory
        project: cluster-project
        location: us-east1
        cluster-name: my-cluster
      aws:
        # optional SA account id/secret because you can use ambiant credentials or aws-iam-controller
        key-id: ASDFGHJK
        key-secret: QWERTYUIOP
        # mandatory
        region: us-east-1
        account-id: 1234567890
        cluster-name: my-cluster
      azure:
        # [...]
```

The users would have to map the cloud's service account to a kubernetes user/group and create a rolebinding to allow this group to use the `TokenReview` API. This step is cloud-specific and can be straightforward (as in GKE) or complex (hello AWS).

I'll add those details to the RFD so they can be reviewed at the same time by the auditors.

Author: done
